src/llama-graph.cpp

```diff
@@ -879,13 +879,24 @@ ggml_tensor * llm_graph_context::build_moe_ffn(
         selection_probs = logits;
     }
 
+    if (arch == LLM_ARCH_GROVEMOE) {
+        selection_probs = ggml_sigmoid(ctx0, logits); // [n_expert, n_tokens]
+        cb(selection_probs, "ffn_moe_probs_biased", il);
+    }
+
     // select experts
     ggml_tensor * selected_experts = ggml_top_k(ctx0, selection_probs, n_expert_used); // [n_expert_used, n_tokens]
     cb(selected_experts->src[0], "ffn_moe_argsort", il);
     cb(selected_experts, "ffn_moe_topk", il);
 
-    ggml_tensor * weights = ggml_get_rows(ctx0,
-            ggml_reshape_3d(ctx0, probs, 1, n_expert, n_tokens), selected_experts); // [1, n_expert_used, n_tokens]
+    if (arch == LLM_ARCH_GROVEMOE && n_expert != hparams.n_expert) {
+        selected_experts = ggml_div_scalar_i32(ctx0, selected_experts, hparams.n_group_experts);
```

Comment:
Given `A / n == A * (1/n)`, so can we do this?

```suggestion
        selected_experts = ggml_scale(ctx0, selected_experts, 1.0f / float(hparams.n_group_experts));
```

Reply:
No, `selected_experts` is an `I32`.